maps/sets.go

```diff
@@ -0,0 +1,149 @@
+package maps
+
+import (
+	"sync"
+
+	"golang.org/x/exp/maps"
+)
+
+// Set is a concurrency safe wrapper around a map to ensure uniqueness through
+// a collection.
+// While the underlying map can be manually manipulated, read and
+// mutation should be done with through methods to ensure
+// concurrency safety.
+type Set[T comparable] struct {
+	mu sync.RWMutex
+
+	Set map[T]struct{}
+}
+
+func (s *Set[T]) Add(vs ...T) {
+	if len(vs) == 0 {
+		return
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.Set == nil {
+		s.Set = make(map[T]struct{}, len(vs))
+	}
+
+	for _, v := range vs {
+		s.Set[v] = struct{}{}
+	}
+}
+
+func (s *Set[T]) Has(v T) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if len(s.Set) == 0 {
+		return false
+	}
+
+	_, ok := s.Set[v]
+
+	return ok
+}
+
+func (s *Set[T]) Keys() []T {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if len(s.Set) == 0 {
+		return nil
+	}
+
+	return Keys(s.Set)
+}
+
+func (s *Set[T]) Delete(vs ...T) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, v := range vs {
+		delete(s.Set, v)
+	}
+}
+
+// IndexedSet is a Set where the key of the map is extracted from
+// Fn.
+// Adding values with the same key will replace the previous value
+type IndexedSet[K comparable, T any] struct {
```

Comment:
what kind of usecase would you use this in?

All the usecases i can think about are taken into account by https://github.com/upfluence/pkg/blob/master/slices/index_by.go#L4

Reply:
Where you will have used index_by several times in a row while scrolling.
Let's take entities A and B, where A has B as a foreign key.  You want to list everything a subset of A based an a value of B and return the related B.
Without the possibility of a join, you will have to fetch both A and B. Filter A by B. Repeat until you have everything you need.
1. Fetch A 
2. Fetch related B -> read in Set first
3. Filter A with B filter
4. Index B in Set to avoid refetch.
5. Return A and B (Set.Values)

It would be lot simpler if we could fetch all of A in one shot but that's not always the case

Reply:
I am not sure to understand your example, lets chat about this tomorrow